rust/primary/src/garbage_collector.rs

```diff
@@ -0,0 +1,68 @@
+use crate::messages::Certificate;
+use crate::primary::PrimaryWorkerMessage;
+use bytes::Bytes;
+use config::Committee;
+use crypto::PublicKey;
+use network::SimpleSender;
+use std::net::SocketAddr;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+use tokio::sync::mpsc::Receiver;
+
+/// Receives the highest round reached by consensus and update it for all tasks.
+pub struct GarbageCollector {
+    /// The current consensus round (used for cleanup).
+    consensus_round: Arc<AtomicU64>,
+    /// Receives the ordered certificates from consensus.
+    rx_consensus: Receiver<Certificate>,
+    /// The network addresses of our workers.
+    addresses: Vec<SocketAddr>,
+    /// A network sender to notify our workers of cleanup events.
+    network: SimpleSender,
+}
+
+impl GarbageCollector {
+    pub fn spawn(
+        name: &PublicKey,
+        committee: &Committee,
+        consensus_round: Arc<AtomicU64>,
+        rx_consensus: Receiver<Certificate>,
+    ) {
+        let addresses = committee
+            .our_workers(name)
+            .expect("Our public key or worker id is not in the committee")
+            .iter()
+            .map(|x| x.primary_to_worker)
+            .collect();
+
+        tokio::spawn(async move {
+            Self {
+                consensus_round,
+                rx_consensus,
+                addresses,
+                network: SimpleSender::new(),
+            }
+            .run()
+            .await;
+        });
+    }
+
+    async fn run(&mut self) {
+        while let Some(certificate) = self.rx_consensus.recv().await {
+            // TODO [issue #9]: Re-include batch digests that have not been sequenced into our next block.
```

Comment:
how does the core know which blocks to cleanup and reinject and which are committed?

Reply:
The core knows which blocks to clean through the shared variable `consensus_round`:
https://github.com/facebookresearch/narwhal/blob/1ff0d968e8a01a6c97015497baf5fb996c38fa21/rust/primary/src/core.rs#L400-L408

It however does not know how to re-inject batch digests that have not been sequences (there is an open issue for that)